<H2>Description</H2>

<p>This track depicts RNA-Seq data from human 
<a target="_BLANK" href="http://hgwdev.cse.ucsc.edu/cgi-bin/hgEncodeVocab?type=rnaExtract">
RNA</a> between the sizes of 20-200 nt. 
As part of the ENCODE Consortia, Cold Spring Harbor Laboratories isolated 
<a target="_BLANK" href="http://hgwdev.cse.ucsc.edu/cgi-bin/hgEncodeVocab?type=localization">
tissues or sub cellular compartments</a> 
from
<a target="_BLANK" href="http://hgwdev.cse.ucsc.edu/cgi-bin/hgEncodeVocab?type=cellType">
ENCODE cell lines</a>
The overall goal of the ENCODE project is to identify and characterize all functional
elements in the sequence of the human genome.
</p>

<p>This cloning protocol generates directional libraries that are read from the
5&#8242; ends of the inserts, which should largely correspond to the 5&#8242;
ends of the mature RNAs.  The libraries were sequenced on a Solexa platform for
a total of 36, 50 or 76 cycles however the reads undergo post-processing
resulting in trimming of their 3&#8242; ends.  Consequently, the mapped read
lengths are variable.
</p>


<h2>Display Conventions and Configuration</h2>

<p>To show only selected subtracks, uncheck the boxes next to the tracks that
you wish to hide.</p>

<p>Color differences among the views are arbitrary. They provide a
visual cue for
distinguishing between the different cell types and compartments.</p>

<dl>
<dt><i>Transfrags</i>
</dt><dd>Small RNA reads were assembled into "Transfrags" by merging reads with one or more overlapping nucleotides. In order to minimize ambiguity from reads that have the potential to map to multiple genomic loci, only the uniquely mapping reads were used to generate Transfrags. The BED6+ format of the transfrag files are created from "intervals-to-contigs" Galaxy tool written by <A HREF="mailto:&#103;&#111;r&#100;&#111;n&#64;&#99;&#115;h&#108;.e&#100;&#117;">Assaf Gordon</a> in the Hannon lab at CSHL. A complete description of the columns in this format can be found <a HREF="../ENCODE/transfragBed6CshlDescription.html">here</a>.
The Transfrags view includes all transfrags before filtering.
</dd><dt><i>Raw Signals</i>
</dt><dd>The Raw Signal view shows the density of mapped reads on the plus and minus strands (wiggle format).
</dd>
<dt><i>Alignments</i>
</dt><dd>The Alignments view shows individual reads mapped from pooled biological replicates to the genome and indicates where
bases may mismatch. Every mapped read is displayed, i.e. uncollapsed.  The track may be set so that the color of each alignment denotes the mapping frequency of the corresponding read:</br>
<font color="#FF00FF">pink</font> for reads mapping to a single genomic location, </br>
<font color="#CC00FF"> pinkish-purple</font> for reads mapping to exactly two genomic locations and </br>
<font color="#800080">purple</font> for reads mapping to more than two possible locations.</br>
  The alignment file follows the standard SAM format of Bowtie output with the following additions: the mapping frequency for each read is indicated in the BAM file using the tag 'YG' and the alignment color is specified in RGB space using the tag 'YC'.  The 'XA' tag is Bowtie tag that means the aligned read belongs to <A HREF=http://bowtie-bio.sourceforge.net/manual.shtml#strata>strata</A>. See the <a href="http://bowtie-bio.sourceforge.net/manual.shtml#sam-bowtie-output" target=_blank>Bowtie Manual</a> for more information about the SAM Bowtie output (including other tags) and the <a href="http://samtools.sourceforge.net/SAM1.pdf#page=4" target=_blank>SAM Format Specification</a> for more information on the SAM/BAM file format. Mapping quality is not available for this track and so, in accordance with the SAM Format Specification, a score of 255 is used. Also, as described in the Methods section below, the sequence quality scores are not available and so are all displayed as 40.
</dd>

<!--Not releaseing in release 2
<dt><i>Contigs</i>
</dt><dd> The Contigs are in .BED format files representing blocks of overlapping mapped reads from pooled biological replicates.
</dd>
</dl>
-->

<H2>Methods</H2>

<p>Small RNA between 20-200 nt was Ribominus-ed<sup>TM</sup>;
 treated according to the
manufacturer's protocol (Invitrogen).
The RNA was treated with Tobacco Alkaline Pyrophosphatase to eliminate any
5&#8242; structures that would inhibit cloning.  
The RNA is then "A-tailed" using Poly-A Polymerase.  A 5&#8242; linker is ligated using T4 RNA ligase and anchored oligo-dT is used to prime a reverse transcriptase reaction.  The libraries are gel purified and clustered for an Illumina GAIIx run. Sequence is obtained from the 5&#8242; end of the inserts for a total of 36 cycles. For a detailed protocol see: </br>###[Small RNA Cloning Protocol.pdf].
</p>

<p>The Illumina reads were initially trimmed to discard any bases following a quality score less than or equal to 20 and converted into FASTA format, thereby discarding quality information for the rest of the pipeline. As a result, the sequence quality scores in the BAM output are all displayed as "40" to indicate no quality information. The read lengths may exceed the insert sizes and consequently introduce
3&#8242; adapter sequence into the 3&#8242; end of the reads.  The 3&#8242;
sequencing adapter was removed from the reads using a custom clipper program (available at <a href="http://hannonlab.cshl.edu/fastx_toolkit">http://hannonlab.cshl.edu/fastx_toolkit/</a>), which aligned the adapter sequence to the short-reads using up to 2 mismatches and no indels.  Regions that aligned were "clipped" off from the read. Terminal C nucleotides introduced at the 3' end of the RNA via the cloning procedure are also trimmed.  The trimmed portions were collapsed into identical reads, their count noted and aligned to the human genome (version hg19, using the gender build appropriate to the sample in question - <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/referenceSequences/female.hg19.fa.gz">female</a>/<a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/referenceSequences/male.hg19.fa.gz">male</a>) using Bowtie  (Langmead B. et al). The alignment parameter allowed 0, 1, or 2 mismatches iteratively. We report reads that mapped 20 or fewer times.</p>

<p><em><strong>Discrepancies between hg18 and hg19 versions of CSHL small RNA data:</em></strong> The alignment pipeline for the CSHL small RNA data was updated upon the release of the human genome version hg19, resulting in a few noteworthy discrepancies with the hg18 dataset.  First, mapping was conducted with the open-source Bowtie algorithm (<a href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a>) rather than the custom NexAlign software.  As each algorithm uses different strategies to perform alignments, the mapping results may vary even in genomic regions that do not differ between builds.  The read processing pipeline also varies slightly, in that we no longer retain information regarding whether a read was 'clipped' off adapter sequence. 
</p>

<p>Data from the Gingeras and Guigo labs were mapped with STAR. For a description of STAR, the source code and mapping parameters used see: <a href="http://gingeraslab.cshl.edu/STAR/">http://gingeraslab.cshl.edu/STAR/<a>. Reads mapping 10 times or less are reported in the Raw Signal and Alignment files.
</p>


<H2>Verification</H2>

<P>Comparison of referential data generated from 8 individual sequencing lanes 
(Illumina technology).   The mapped data were also visually inspected to verify the majority of the reads were mapping the 5&#8242; ends of annotated small RNA classes.
</P>

<H2>Credits</H2>
Hannon lab members: Katalin Fejes-Toth, Vihra Sotirova, Gordon  Assaf, Jon Preall</P>
<P>Gingeras and Guigo laboratories: Carrie A. Davis, Lei-Hoon See, Wei Lin. </P>

Contacts: 
<DL>
<DD><A HREF="mailto:&#106;&#112;r&#101;&#97;&#108;&#108;&#64;&#99;&#115;&#104;l.
&#101;&#100;&#117;">Jonathan Preall</A> (Generation 0 Data from Hannon Lab) 
<!-- above address is jpreall at cshl.edu -->
</DD>
<DD>
<A HREF="mailto:&#100;a&#118;&#105;s&#99;&#64;&#99;sh&#108;.&#101;&#100;&#117;">Carrie Davis</A> (experimental)
<!-- above address is davisc at cshl.edu -->
</DD>
<DD>
<A HREF="mailto:&#100;&#111;&#98;i&#110;&#64;&#99;&#115;&#104;l.&#101;d&#117;">Alex Dobin</A> (computational)
<!-- above address is dobin at cshl.edu -->
</DD>
<DD>
<A HREF="mailto:&#115;&#99;&#104;&#108;&#101;si&#110;&#64;&#99;&#115;&#104;l.e&#100;&#117;">Felix
Schlesinger</A> (computational)
<!--above address is schlesin at cshl.edu -->
</DD>
<DD>
<A HREF="mailto:&#103;&#105;&#110;&#103;e&#114;a&#115;&#64;&#99;&#115;&#104;&#108;.&#101;d&#117;">Tom Gingeras</A> (primary investigator)
<!--above address is gingeras at cshl.edu -->
</P>
</DD>
</DL>

<H2> Data Release Policy </H2>

<P>Data users may freely use ENCODE data, but may not, without prior 
consent, submit publications that use an unpublished ENCODE dataset until 
nine months following the release of the dataset.  This date is listed in 
the <EM>Restricted Until</EM> column, above.  The full data release policy 
for ENCODE is available 
<A HREF="../ENCODE/terms.html" TARGET=_BLANK>here</A>.</P> 
